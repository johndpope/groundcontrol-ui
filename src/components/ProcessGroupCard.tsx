// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import graphql from "babel-plugin-relay/macro";
import { Link } from "found";
import React from "react";
import Moment from "react-moment";
import { createFragmentContainer } from "react-relay";
import {
  Button,
  Card,
  Icon,
 } from "semantic-ui-react";

import { ProcessGroupCard_item } from "./__generated__/ProcessGroupCard_item.graphql";

import ProcessTable from "./ProcessTable";
import { IProps as IProcessTableRowProps } from "./ProcessTableRow";

import "./ProcessGroupCard.css";

const dateFormat = "L LTS";

export interface IProps {
  item: ProcessGroupCard_item;
  onStartGroup: (values: IProps) => any;
  onStopGroup: (values: IProps) => any;
  onStartProcess: (values: IProcessTableRowProps) => any;
  onStopProcess: (values: IProcessTableRowProps) => any;
}

export function ProcessGroupCard(props: IProps) {
  const {
    item: {
      createdAt,
      status,
      processes,
      task,
      task: {
        workspace,
      },
    },
    onStartGroup,
    onStopGroup,
  } = props;
  const processNodes = processes.edges.map(({ node }) => node);
  const buttons: JSX.Element[] = [];
  const handleStart = () => onStartGroup({ ...props });
  const handleStop = () => onStopGroup({ ...props });

  switch (status) {
  case "RUNNING":
    buttons.push((
      <Button
        key="stop"
        color="teal"
        floated="right"
        icon={true}
        labelPosition="left"
        onClick={handleStop}
      >
        <Icon name="stop" />
        Stop Group
      </Button>
    ));
    break;
  case "DONE":
  case "FAILED":
    buttons.push((
      <Button
        key="start"
        color="teal"
        floated="right"
        icon={true}
        labelPosition="left"
        onClick={handleStart}
      >
        <Icon name="play" />
        Start Group
      </Button>
    ));
    break;
  }

  return (
    <Card
      className="ProcessGroupCard"
      fluid={true}
    >
      <Card.Content>
        <Card.Header>
          {buttons}
          <Link to={`/workspaces/${workspace.slug}`}>
            {workspace.name}
          </Link> / {task.name}
        </Card.Header>
        <Card.Meta>
          <Moment format={dateFormat}>{createdAt}</Moment>
        </Card.Meta>
      </Card.Content>
      <ProcessTable
        {...props}
        items={processNodes}
      />
    </Card>
  );
}

export default createFragmentContainer(ProcessGroupCard, graphql`
  fragment ProcessGroupCard_item on ProcessGroup {
    id
    createdAt
    status
    task {
      name
      workspace {
        slug
        name
      }
    }
    processes {
      edges {
        node {
          ...ProcessTable_items
        }
      }
    }
  }`,
);
