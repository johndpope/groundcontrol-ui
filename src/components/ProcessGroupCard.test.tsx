// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import { ProcessGroupCard_item } from "./__generated__/ProcessGroupCard_item.graphql";

import { mockQueryPropAttrs } from "../testing/relay";

import { IProps, ProcessGroupCard } from "./ProcessGroupCard";

const item: ProcessGroupCard_item = {
  ...mockQueryPropAttrs(),
  createdAt: "createdAt",
  id: "id",
  processes: {
    edges: [{
      node: {
        ...mockQueryPropAttrs(),
      },
    }, {
      node: {
        ...mockQueryPropAttrs(),
      },
    }],
  },
  status: "RUNNING",
  task: {
    name: "taskName",
    workspace: {
      name: "workspaceName",
      slug: "workspaceSlug",
    },
  },
};

const props = {
  item,
  onStartGroup: jest.fn(),
  onStartProcess: jest.fn(),
  onStopGroup: jest.fn(),
  onStopProcess: jest.fn(),
};

beforeEach(() => {
  mocked(props.onStartGroup).mockClear();
  mocked(props.onStopGroup).mockClear();
});

describe("<ProcessGroupCard />", () => {

  it("renders correctly", () => {
    const wrapper = shallow(<ProcessGroupCard {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("show a stop group button if the process group is running", () => {
    const wrapper = shallow(
      <ProcessGroupCard
        {...{...props, item: { ...item, status: "RUNNING" } }}
      />,
    );
    expect(wrapper.find("Button[icon='stop']")).toHaveLength(1);
  });

  it("show a start group button if the process group is done", () => {
    const wrapper = shallow(
      <ProcessGroupCard
        {...{...props, item: { ...item, status: "DONE" } }}
      />,
    );
    expect(wrapper.find("Button[icon='play']")).toHaveLength(1);
  });

  it("show a start group button if the process group failed", () => {
    const wrapper = shallow(
      <ProcessGroupCard
        {...{...props, item: { ...item, status: "FAILED" } }}
      />,
    );
    expect(wrapper.find("Button[icon='play']")).toHaveLength(1);
  });

  it("triggers onStopGroup when the stop group button is clicked", () => {
    const wrapper = shallow(
      <ProcessGroupCard
        {...{...props, item: { ...item, status: "RUNNING" } }}
      />,
    );
    wrapper.find("Button[icon='stop']").simulate("click");
    expect(props.onStopGroup).toBeCalledTimes(1);
    expect(props.onStopGroup).toBeCalledWith(props);
  });

  it("triggers onStartGroup when the start group button is clicked", () => {
    const doneProps: IProps = {...props, item: { ...item, status: "DONE" } };
    const wrapper = shallow(<ProcessGroupCard {...doneProps} />);
    wrapper.find("Button[icon='play']").simulate("click");
    expect(doneProps.onStartGroup).toBeCalledTimes(1);
    expect(doneProps.onStartGroup).toBeCalledWith(doneProps);
  });

});
