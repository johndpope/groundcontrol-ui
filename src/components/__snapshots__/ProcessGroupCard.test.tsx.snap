// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProcessGroupCard /> renders correctly 1`] = `
<Card
  className="ProcessGroupCard"
  fluid={true}
>
  <CardContent>
    <CardHeader>
      <Button
        as="button"
        color="teal"
        content="Stop Group"
        floated="right"
        icon="stop"
        key="stop"
        onClick={[Function]}
      />
      <Link
        to="/workspaces/workspaceSlug"
      >
        workspaceName
      </Link>
       / 
      taskName
    </CardHeader>
    <CardMeta>
      <Moment
        format="L LTS"
      >
        createdAt
      </Moment>
    </CardMeta>
  </CardContent>
  <ForwardRef(Relay(ProcessTable))
    item={
      Object {
        " $fragmentRefs": null,
        " $refType": null,
        "createdAt": "createdAt",
        "id": "id",
        "processes": Object {
          "edges": Array [
            Object {
              "node": Object {
                " $fragmentRefs": null,
                " $refType": null,
              },
            },
            Object {
              "node": Object {
                " $fragmentRefs": null,
                " $refType": null,
              },
            },
          ],
        },
        "status": "RUNNING",
        "task": Object {
          "name": "taskName",
          "workspace": Object {
            "name": "workspaceName",
            "slug": "workspaceSlug",
          },
        },
      }
    }
    items={
      Array [
        Object {
          " $fragmentRefs": null,
          " $refType": null,
        },
        Object {
          " $fragmentRefs": null,
          " $refType": null,
        },
      ]
    }
    onStartGroup={[MockFunction]}
    onStartProcess={[MockFunction]}
    onStopGroup={[MockFunction]}
    onStopProcess={[MockFunction]}
  />
</Card>
`;
